vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/step/DescendantOrSelfPathStepUnnesting.java

```diff
@@ -136,8 +156,23 @@ protected boolean stepNodeTree(TaggedValuePointable tvpInput, int level, IPointa
                 // Check current node
                 if (indexSequence.get(level) == returnSequence.get(level)) {
                     returnSequence.set(level, returnSequence.get(level) + 1);
-                    setNodeToResult(tvpItem, result);
-                    return true;
+                    if (isfilter) {
```

Comment:
The logic here is not correct. The process adds nodes to the result that do not match the filter.